BHoM/BHoMAnalytics_Toolkit
# Obtain Project ID for analytics through more direct means

### NOTE: Depends on 
https://github.com/BHoM/BHoM_UI/pull/425

   
### Issues addressed by this PR
Fixes #34 


### Test files
Not Available - follow the testing steps below.

**To compile the code**

1. Run an appropriate alpha installer
2. Compile BHoM_UI first (see [here](https://github.com/BHoM/BHoM_UI/pull/425)
3. Compile BHoMAnalytics_Toolkit
4. Run the tests below

**To test these Pull Requests**
 1. For a chosen UI (ideally this ought to be tested in both Grasshopper and Excel as a minimum, with Dynamo given consideration if there's time/appetite), open a new script (for the purposes of this explanation, Grasshopper terms will be used, and the equivalent term should be applied for Excel when testing there). No pop up should appear.
 6. Place a random component on the canvas that is **NOT** a BHoM component. No pop up should appear.
 7. Place a component on the canvas that **IS** a BHoM component - but do **NOT** provide it any inputs. No pop up should appear.
 8. Provide that component some inputs so that it runs. A pop up should appear (see screenshot below for what is expected).
 9. Click the `Non-Project` option - the pop up should disappear.
 10. Place another component on the canvas that **IS** a BHoM component - but do **NOT** provide it any inputs. No pop up should appear.
 11. Provide inputs to that new component so that it runs. No pop up should appear (because we have already provided the non-project designation).
 12. Close the script entirely.
 13. Check the logs file for that session and see that the final component has the `ProjectID` of `Non-Project`.
 14. Repeat steps 1-4
 15. Click the `Project` option - the pop up should expand (see second screenshot) asking for project number.
 16. Provide a project number and click `Confirm` - the pop up should disappear.
 17. Repeat steps 6-8
 18. Check the logs file for that session and see that the final component has the `ProjectID` of the value you just provided.
 19. Repeat for additional UIs as appropriate.

**Grasshopper only**

Grasshopper has a template file which provides users an option to provide a project ID by default. This test will confirm that if the user uses that option FIRST, they do not get the pop-up.

1. Open a new Grasshopper script.
2. Please a random component on the canvas that is **NOT** a BHoM component - this is to trigger the template file loading. No pop up should appear.
3. Provide a project ID on the input provided by the template.
4. Place a component on the canvas that **IS** a BHoM component - but do **NOT** provide it any inputs. No pop up should appear.
5. Provide inputs to that new component so that it runs. No pop up should appear.
6. Close the script entirely.
7. Check the logs file for that session and see that the final component has the `ProjectID` of the value you just provided.

**Checking lack of pop-up when Analytics is not installer**

These tests will confirm that non-analytic based installers (such as the betas) do not get affected by the changes on BHoM_UI.

 1. Run an appropriate BHoM installer from [here](https://bhom.xyz)
 2. Compile BHoM_UI - do **NOT** compile BHoMAnalytics_Toolkit
 3. Run a test suitable for any UI - verify that no pop-up appears when using BHoM components.

**Initial pop-up window**
![image](https://user-images.githubusercontent.com/18049174/184323704-8667f435-312b-4bda-af06-885a38919e47.png)

**Pop up window AFTER selecting the Project Option**
![image](https://user-images.githubusercontent.com/18049174/184323770-fbf0e12d-8f47-4fc8-b429-a381c9ae711d.png)

**Several things to note**

 - I am not a graphics designer - this pop-up does not look pretty. I am hoping that @alexissantella and others with graphical capabilities can assist in improving this in a later PR.
 - The wording may also require improvement - I am not a wordsmith outside my published poetry, but poetry isn't appropriate here either so I've gone with wording that is small, clear, and concise for now. Again, I am hoping that @alexissantella and others will be able to support and assist in updating the wording in a later PR.
 - There are later phases of work in relation to this - this gets the framework in which should then allow @alexissantella and others to support in expanding this functionality in future PRs.
